fix: uneven load among clickhouse shards caused by retry error mechanism

nir3c opened the pull request:
## Description


fix uneven load among clickhouse shards caused by the retry error mechanism (related to issues #325 and #322 )

The issue happens when ChProxy sends the HTTP call to one of the shards with a retry mechanism.

When the response from the shard is 502 the [scope's host swaps](https://github.com/ContentSquare/chproxy/blob/6fb82516c6b2066a3a4cd271f22ae741e75b551d/proxy.go#L251) with another one to send by it, which causes not to decreasing the initial host's counter and the new host's counter to decrease instead (when we finally [decrementing the scope](https://github.com/ContentSquare/chproxy/blob/6fb82516c6b2066a3a4cd271f22ae741e75b551d/proxy.go#L107)), which could lead to a situation when the new host's counter is 0 to be set to MaxUint32 value (as `0 - 1` equals to the maximum value of uint32).
once this situation happens the new host will not be able to receive any new requests (as its counter value will be always greater than all other hosts' counter, and the [balancing function logic for getting the new host](https://github.com/ContentSquare/chproxy/blob/6fb82516c6b2066a3a4cd271f22ae741e75b551d/scope.go#L1011) is [based on the host's counter + penalty](https://github.com/ContentSquare/chproxy/blob/6fb82516c6b2066a3a4cd271f22ae741e75b551d/scope.go#L823)) and the host's replica will never be used (same as for the host -> it will always be higher value than all other replicas as we can see in the [getReplica function](https://github.com/ContentSquare/chproxy/blob/6fb82516c6b2066a3a4cd271f22ae741e75b551d/scope.go#L900))


<!--  Please explain the object of this PR and the changes you made.
A reference to a github issue would be appreciated. --> 

## Pull request type

<!-- Please do not submit updates to dependencies unless it fixes an issue. --> 

<!-- Please try to limit your pull request to one type, submit multiple pull requests if needed. --> 

Please check the type of change your PR introduces:
- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):


### Checklist

- [ ] Linter passes correctly
- [ ] Add tests which fail without the change (if possible)
- [x] All tests passing
- [ ] Extended the README / documentation, if necessary

## Does this introduce a breaking change?

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->

- [ ] Yes
- [ ] No

## Further comments

<!-- If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc... -->

nir3c committed:
fix uneven load among clickhouse shards caused by retry error mechanism

sigua-cs commented:
@nir3c thanks for the PR. Please check failed tests

nir3c commented on proxy.go:
```diff
@@ -241,14 +241,22 @@ func executeWithRetry(
 			// comment s.host.dec() line to avoid double increment; issue #322
 			// s.host.dec()
```
@sigua-cs should I remove this comment as I add the decrement in line 255?

mga-chka replied:
yes you should since your PR fix this legacy quick & dirty (and broken) fix.

nir3c replied:
done

mga-chka commented:
Thanks for the PR.
FYI, since your PR is linked to the tricky issues we had with the retry mechanism, we will take a bit of time to review it (since we will need multiple reviewer and because of the holidays it will take more time).

Blokje5 commented on proxy.go:
```diff
@@ -241,14 +241,22 @@ func executeWithRetry(
-
+				currentHost := s.host
+
+				// decrement the current failed host counter and increment the new host
```
It might also a be a good idea to link the PR.

nir3c replied:
done

Blokje5 reviewed (approved):
LGTM! Great find.

mga-chka commented on proxyretry_test.go:
```diff
@@ -154,7 +155,17 @@ func TestQueryWithRetrySuccess(t *testing.T) {
+	assert.Equal(t, 1, int(s.host.load()))
+
+	assert.Equal(t, 0, int(erroredHost.counter.load()))
+	assert.Equal(t, 5, int(erroredHost.penalty))
```
You should use the **penaltySize** variable instead of 5, it would make the test more readable

nir3c replied:
done

mga-chka reviewed (approved):
thanks for the PR, sorry for the long delay (we take a lot of days off in summer in France).
I left a few comments but they're very minor. Please tell me we want to fix them so that we don't wait for an extra 2 weeks before we merge the PR.

Once the PR is merged, we will release a new version ASAP so that you can benefit from your fix.

nir3c committed:
CR changes 
1. add magic number penaltySize for test assert the number of error host's load + penalty result
2. update comments

nir3c commented:
> thanks for the PR, sorry for the long delay (we take a lot of days off in summer in France). I left a few comments but they're very minor. Please tell me we want to fix them so that we don't wait for an extra 2 weeks before we merge the PR.
> 
> Once the PR is merged, we will release a new version ASAP so that you can benefit from your fix.

thanks for reviewing the code, I update the code based on the comments you gave me, feel free review it again

mga-chka commented:
> > thanks for the PR, sorry for the long delay (we take a lot of days off in summer in France). I left a few comments but they're very minor. Please tell me we want to fix them so that we don't wait for an extra 2 weeks before we merge the PR.
> > Once the PR is merged, we will release a new version ASAP so that you can benefit from your fix.
> 
> thanks for reviewing the code, I update the code based on the comments you gave me, feel free review it again

We should release a new version containing your fix on Monday.
